### PR TITLE
VBF ML development for event images

### PR DESCRIPTION
**Pull request by alpakpinar**
Hey @AndreasAlbert, @abdelazizhussein, here is the first version of the ML code I've been working on for training on event images. The code in this PR passes all tests and is formatted with `black`. I still have some features to implement (mostly have to do with checks on training for event image case), but wanted to open this PR in case you want to merge already, or suggest some changes.

This PR has:
- Two models (dense, convolutional) defined separately with their configurations exported to a yml file under `config` directory
- A script to transform pickle output from bucoffea to ROOT trees (only for event images)
- A few updates to the `MultiDatasetSequence` class to cover some edge cases (e.g. files with very low number of events)
- New tests for the `ModelDB` class that is used as an interface between the `yml` config and the training script.

This PR does not have the following yet:
- Validation of training vs validation data for event images by comparing some distributions between the two
- Training analyzer implementation for the training with event images

Let me know if you have any thoughts, comments, thanks!

### CODE REVIEW

File: vbfml/models.py [hunk 1]
```diff
@@ -34,3 +43,48 @@ def sequential_dense_model(
...
+    n_filters_for_conv: "list[int]",
+    pool_size_for_conv: "list[int]",
+    filter_size_for_conv: "list[int]",
+    n_nodes_for_dense: "list[int]",
+    n_layers_for_dense: int,
+    n_classes: int,
+    dropout: float = 0,
+    image_shape: tuple = (40, 20, 1),
+) -> Sequential:
+    model = Sequential()
```
Comment (AndreasAlbert) [inline vbfml/models.py:59]: Add docstiring please

File: vbfml/scripts/pickle_to_root.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import numpy as np
+
+from tqdm import tqdm
+from pprint import pprint
+from vbfml.util import vbfml_path
+
+pjoin = os.path.join
+
+
+def main():
```
Comment (AndreasAlbert) [inline vbfml/scripts/pickle_to_root.py:17]: Add docstring explaining what this script is used for, what the input and outputs are

File: vbfml/scripts/train.py [hunk 3]
```diff
@@ -44,39 +49,25 @@ def cli(ctx, tag):
...
 )
 @click.option("--dropout", default=0.5, required=False, help="Dropout rate.")
-def setup(ctx, learning_rate: float, dropout: float):
+@click.option(
+    "--input-dir",
+    default="./output/2021-10-29_vbfhinv_26Oct21_nanov8",
+    required=False,
+    help="Input directory containing the ROOT files for training and validation.",
+)
+@click.option(
```
Comment (AndreasAlbert) [inline vbfml/scripts/train.py:58]: Instead of having a predefined set of choices here, why not just feed in a configuration file?
```
train.py --model-config /path/to/config.yml
```

Then, we can have entirely separate config files for each type of architecture. In the current proposal, the config file has settings for both convolutional an ddense models which makes no sense. If we split, we can have one common key for all architectures that tells the parser what type of model to build, e.g.:
```
architecture: conv
```

this will also give us flexibility because we can now easily add different types of new architectures

File: vbfml/scripts/train.py [hunk 4]
```diff
@@ -87,38 +78,69 @@ def setup(ctx, learning_rate: float, dropout: float):
         if re.match(dataset_labels["v_qcd_nlo_17"], dataset_info.name):
             dataset_info.n_events = 0.01 * dataset_info.n_events
 
+    # Object containing data for different models
+    # (set of features, dropout rate etc.)
+    # Loaded from the YML configuration file
+    ymlfile = "../config/model_params.yml"
+    db = ModelDB(ymlfile)
```
Comment (AndreasAlbert) [inline vbfml/scripts/train.py:85]: Suggest different name for "ModelDB" since this is not really a data base. It's a configuration file parser.

File: vbfml/scripts/train.py [hunk 5]
```diff
@@ -87,38 +78,69 @@ def setup(ctx, learning_rate: float, dropout: float):
...
 
     # Build model
-    model = sequential_dense_model(
-        n_layers=3,
-        n_nodes=[4, 4, 2],
-        n_features=len(features),
-        n_classes=len(training_sequence.dataset_labels()),
-        dropout=dropout,
-    )
+    if model == "dense":
```
Comment (AndreasAlbert) [inline vbfml/scripts/train.py:121]: let's put all of this in the config file and put the creation code in a factory so that I can just call:
```
model_config = ModelConfiguration("/path/to/file.yml")
model = ModelFactory.build(model_config)
```

(which ModelFactory.build being a class method)

File: vbfml/util.py [hunk 6]
```diff
@@ -1,7 +1,67 @@
...
+
+    def load(self) -> dict:
+        """Loads and returns data from a YAML file."""
+        with open(self.infile, "r") as f:
+            data = yaml.load(f, Loader=yaml.FullLoader)
+            return data
+
+
+@dataclass
+class ModelDB:
```
Comment (AndreasAlbert) [inline vbfml/util.py:28]: please add type hints everywhere

**Review by AndreasAlbert — CHANGES_REQUESTED** [review]

File: vbfml/models.py (hunk 1 above)
Comment (alpakpinar) [reply vbfml/models.py:87]: Done in f8160d7

File: vbfml/scripts/pickle_to_root.py (hunk 2 above)
Comment (alpakpinar) [reply vbfml/scripts/pickle_to_root.py:17]: Done in f8160d7

File: vbfml/scripts/train.py (hunk 3 above)
Comment (alpakpinar) [reply vbfml/scripts/train.py:58]: Thanks, I split the configuration files for each model in 88caa8d

File: vbfml/scripts/train.py (hunk 4 above)
Comment (alpakpinar) [reply vbfml/scripts/train.py]: You're right, that class name is changed to `ModelConfiguration` in 88caa8d

File: vbfml/util.py [hunk 7]
```diff
@@ -1,7 +1,67 @@
...
+pjoin = os.path.join
+
+
+def vbfml_path(path):
+    """Returns the absolute path for the given path."""
+    return pjoin(vbfml.__path__[0], path)
+
+
+@dataclass
+class YamlLoader:
```
[the remark this answers is not in the document]
Comment (alpakpinar) [reply vbfml/util.py:17]: Sorry I'm sometimes obsessed with classes here and there.. Removed in 88caa8d

File: vbfml/util.py (hunk 6 above)
Comment (alpakpinar) [reply vbfml/util.py]: Done in f6998ee & 230ae9e

File: vbfml/scripts/train.py (hunk 5 above)
Comment (alpakpinar) [reply vbfml/scripts/train.py]: This is implemented in 676a7a3

**Comment by alpakpinar** [comment]
Hey @AndreasAlbert, thanks for the comments! I went through them and made updates in the code. Let me know if you want any further changes

**Review by AndreasAlbert — APPROVED** [review]

**Comment by AndreasAlbert** [comment]
Thanks looks good now!
